futurewei-cloud/arktos
# Not to assign ClusterIP for service whose associated network is of External IPAM type

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
modify kube-apiserver to bypass clusterIP assignment for services whose VIP address space is managed by external network provider. It also adds spec.service section to network CRD.

Based on [service IP managed by network provider](https://github.com/futurewei-cloud/arktos/blob/master/docs/design-proposals/multi-tenancy/multi-tenancy-network.md#network-provider-managed-pool), Arktos leaves service ClusterIP as unassigned (eternal network provider's controller should assign them instead), if the associated networks explicitly state the External IPAM type.

This PR temporarily ignores the error that the default network not created (which would be mitigated by tenant controller if the feature of automatic creation of default network for new tenant is enabled). This temporary measure is necessary to keep system not broken before the required dependencies are all in place. However, if service explicitly associates to a network, the network has to exist.

This is part of multi-tenant networking features.

**How to verify on local dev env**:
1. start etcd, kube-api-server (http protocol, port 8080). The kube-apiserver start command is
```kube-apiserver -- --etcd-servers=http://127.0.0.1:2379```
2. ensure network CRD created
```kubectl apply -f pkg/controller/artifacts/crd-network.yaml```
3. creates a network CR of External IPAM type using following yaml
```yaml
apiVersion: "arktos.futurewei.com/v1"
kind: Network
metadata:
  name: test1
spec:
  type: mizar
  vpcID: 12345-mizar
  service:
    ipam: External
```
4. able to create a service associated to test1 network using following yaml, verify that its will has no clusterIP
```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-nginx
  labels:
    run: my-nginx
    arktos.futurewei.com/network: test1 
spec:
  ports:
  - port: 80
    protocol: TCP
  selector:
    run: my-nginx
```
5. verify creating a service associated to an inexistent network using following yaml fails with error "networks.arktos.futurewei.com test2 not found"
```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-nginx-inexistent-net
  labels:
    run: my-nginx
    arktos.futurewei.com/network: test2 
spec:
  ports:
  - port: 80
    protocol: TCP
  selector:
    run: my-nginx
```

**Does this PR introduce a user-facing change?**:
YES
```release-note
1. If a service associates a network by providing arktos.futurewei.com/network label, the named network has to be created already; otherwise, service cannot be created
2. The temporary exception is services not having such network label. This temporary exception will be removed when the dependency features are all enabled.
```
